### ISSUE TITLE
SmtEngine allows to set Options only once. Feature or Bug?

### ISSUE THREAD

**Issue by kfriedberger** [CVC4/CVC4#3055]
We are using CVC4 (latest version, 2d478849d704f44f2a1f584dbc98d3731ca4e6ee) via the JavaBindings.
We use one central instance of `ExprManager` and several `SmtEngines`.
From the user's point of view the each instance of a SmtEngine is independent (i.e. it has a separate stack for PUSH/POP/ASSERT) and all API calls on SmtEngines are directed towards different instances.
However, the following snippet throws a `RuntimeException`, because it `cannot change option 'produce-models' after final initialization`. 

<pre>
    ExprManager em = new ExprManager();
    Expr x = em.mkVar("x", em.integerType());
    Expr y = em.mkVar("y", em.integerType());

    SmtEngine smt0 = new SmtEngine(em);
    smt0.assertFormula(em.mkConst(true));

    SmtEngine smt = new SmtEngine(em);
    smt.setOption("produce-models", new SExpr(true));
    smt.assertFormula(em.mkExpr(Kind.LT, x, y));
    Result res = smt.checkSat();
    System.out.println("x =  " + smt.getValue(x));
</pre>

Is this really intended behaviour?
Does CVC4 share the set of options over all instances of SmtEngines? 
Are the options given to the central ExprManager, or are there some `central` options that can be set once and others that can be set several times independently?

We might limit our application to setting the same options (especially: enable model-production) for all SmtEngines, if this helps finding a solution. However, we need to create and use several SmtEngines at different points in time.

**Comment by 4tXJ7f** [comment]
Currently, the options are associated with `ExprManager`s and not the `SmtEngine`s, so if you want to set different options, you have to create different `ExprManager`s. In the long run, we are planning to make this less confusing. When you have multiple `ExprManager`s, please note that you cannot mix `Expr`s from different `ExprManager`s and you have to use the associated `ExprManager` to create `Expr`s that you pass to an `SmtEngine`. You can use `Expr::exportTo()` to export an expression to a different `ExprManager`.

I hope this helps, please let me know if you have any other questions.

**Comment by kfriedberger** [comment]
Thanks for your explanation. That answers a lot of points.

Just for clarification, there is a further question:
Is is possible to change an option later (for the same instance of an ExprManager)?

**Comment by 4tXJ7f** [comment]
No, after the first call to `assertFormula`, the options of the `ExprManager` associated with the `SmtEngine` generally can't be changed anymore. This is unlikely to change because certain options (such as `:produce-models`) for example change the kind of preprocessing that we can do. If such an option were changed at a later time, we would have to detect the situation, reset the solver to an earlier point, and replay the steps with the new options. This would be difficult and not very efficient.

I hope that this makes sense and you can work around this limitation. If you have a concrete option that you would like to change on-the-fly we could potentially look into supporting that but in most cases it will unfortunately not be possible.

**Comment by kfriedberger** [comment]
We can work with unmodifiable options, that is no problem.

I am currently playing around with CVC4 and there is also a maybe unexpected behavior when creating new variables:
<pre>
    ExprManager em = new ExprManager();
    SmtEngine smt = new SmtEngine(em);

    Expr x1 = em.mkVar("x", em.integerType());
    Expr x2 = em.mkVar("x", em.integerType());
    Expr formula = em.mkExpr(Kind.DISTINCT, x1, x2);

    smt.assertFormula(formula);
    System.out.println(String.format("Satisfiable: " + smt.checkSat()));
</pre>
It seems that the two variables `x1` and `x2` are distinct variables, even if they have the same name.
Is there an option or API method to avoid such behavior, e.g., such that `mkVar` returns always the same instance of the variable for the same name and type? This behaviour also is visible for `mkFunctionType`.

Additionally, this behaviour would imply that exporting a variable (via Expr::export) into a second ExprManager and creating a new variable with the same name leads to two distinct variables. This makes it impossible to build expressions with two separate ExprManagers and then merge them together, because they will always be different.

**Comment by 4tXJ7f** [comment]
Sorry, I must've missed your last message (please feel free to ping us again if you don't get an answer in a reasonable amount of time).

> It seems that the two variables `x1` and `x2` are distinct variables, even if they have the same name.

You are right. This is by design and the behavior cannot be changed as far as I know. However, you can quite easily avoid the issue by keeping track of the variables that you've created in an `std::string` (variable names) to `Expr` (variable expressions) map. When you request a variable with the same name as an earlier variable, you can return the `Expr` stored in the map. We do something similar in CVC4's parser.

> Additionally, this behaviour would imply that exporting a variable (via Expr::export) into a second ExprManager and creating a new variable with the same name leads to two distinct variables.

If I am understanding correctly, this should actually be possible if you do the following:

1. Create a variable in one `ExprManager`
2. Export the `Expr` for the variable to the second `ExprManager`
3. ??? (do some operations with two solvers)
4. Export the expressions that you want to merge from the second `ExprManager` to the first one

It is important that you keep the same `ExprManagerMapCollection`s for all calls to `exportTo`. This datastructure keeps track of mappings between variables in different `ExprManager`s.

I hope this helps! Please let us know if you have any other questions.

**Comment by kfriedberger** [comment]
Having a user-defined mapping from String to Expr is possible for us. We already use that when having one ExprManager. We will try to extend our implementation to synchronize the mapping of already created variables whenever we export expression from one ExprManager to another one.

For the export, we will use your class `ExprManagerMapCollection`.
How do you guarantee that an Expr exported from `A` to `B` and then from `B` to `C` is identical to an Expr exported directly from `A` to `C`?
Is this class intended for multi-directional usage, i.e., should we use one (central) mapping when translating between three different ExprManagers `A`, `B`, `C` or six instances (one for each direction `A->B`, `A->C`, `B->A`, `B->C`, `C->A`, `C->B`)?

**Comment by 4tXJ7f** [comment]
> How do you guarantee that an Expr exported from `A` to `B` and then from `B` to `C` is identical to an Expr exported directly from `A` to `C`?

We don't ;) This is a bit tricky to do since when `exportTo` gets called, we don't really know about `ExprManager`s other than the two involved. I guess we could somehow find a "mapping path" given the mappings between `ExprManager`s but this would require some changes. Is it possible to design your system such that the mappings are always between a central `ExprManager` and others? E.g. only between `A<->B` and `A<->C` but never `B<->C`?

**Comment by kfriedberger** [comment]
The export `A->B->C` and `A->C` is just a rare case for us and we will try to avoid it.

Further questions:
Is the `ExprManagerMapCollection` bi-directional and should be used for `A->B` and also `B->A`?
Or do we need two different mappings here?
Would there be any way to pre-fill the ExprManagerMapCollection with some data, e.g., directly insert a new mapping for two expressions from two different ExprManagers?

**Comment by 4tXJ7f** [comment]
> Is the ExprManagerMapCollection bi-directional and should be used for A->B and also B->A?

Yes, the `ExprManagerMapCollection` is bi-directional, so for exports from A to B and B to A, you'd pass the same instance.

> Would there be any way to pre-fill the ExprManagerMapCollection with some data, e.g., directly insert a new mapping for two expressions from two different ExprManagers?

The easiest way to "pre-fill" it would be to just export the expressions that you want in there.

Hope this helps!
